examples/text_classification/iterable_train.py

```diff
@@ -0,0 +1,157 @@
+import os
+import logging
+import argparse
+
+import torch
+import sys
+import io
+
+from torchtext.datasets import text_classification
+from torch.utils.data import DataLoader
+
+from model import TextSentiment
+
+from torchtext.vocab import build_vocab_from_iterator
+from torchtext.data.utils import ngrams_iterator
+from torchtext.data.utils import get_tokenizer
+from torchtext.data.utils import build_iterable_dataset_from_iterator
+from torchtext.utils import unicode_csv_reader
+from torchtext.vocab import build_vocab_from_iterator
+
+from tqdm import tqdm
+
+
+def generate_batch(batch):
+
+    def generate_offsets(data_batch):
+        offsets = [0]
+        for entry in data_batch:
+            offsets.append(offsets[-1] + len(entry))
+        offsets = torch.tensor(offsets[:-1])
+        return offsets
+
+    cls = torch.tensor([entry[0] for entry in batch])
+    text = [entry[1] for entry in batch]
+    offsets = generate_offsets(text)
+    text = torch.cat(text)
+    return text, offsets, cls
+
+
+def train(lr_, num_epoch, data_):
+    data = DataLoader(data_, batch_size=batch_size,
+                      collate_fn=generate_batch, num_workers=args.num_workers, pin_memory=True)
+    optimizer = torch.optim.SGD(model.parameters(), lr=lr_)
+    scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer, args.lr_gamma)
+    with tqdm(unit_scale=0, unit='lines', total=num_lines) as t:
+        avg_loss = 0.0
+        for i, (text, offsets, cls) in enumerate(data):
+            optimizer.zero_grad()
+            text, offsets, cls = text.to(device), offsets.to(device), cls.to(device)
+            output = model(text, offsets)
+            loss = criterion(output, cls)
+            loss.backward()
+            avg_loss += loss.item()
+            optimizer.step()
+            if i % (16 * batch_size) == 0:
+                scheduler.step()
+                avg_loss = avg_loss / (16 * batch_size)
+                avg_loss = 0
+                t.set_description("lr: {:9.3f} loss: {:9.3f}".format(scheduler.get_lr()[0], loss))
+            t.update(batch_size)
+
+
+def test(data_):
+    data = DataLoader(data_, batch_size=batch_size, collate_fn=generate_batch)
+    total_accuracy = []
+    for text, offsets, cls in data:
+        text, offsets, cls = text.to(device), offsets.to(device), cls.to(device)
+        with torch.no_grad():
+            output = model(text, offsets)
+            accuracy = (output.argmax(1) == cls).float().mean().item()
+            total_accuracy.append(accuracy)
+    print("Test - Accuracy: {}".format(sum(total_accuracy) / len(total_accuracy)))
+
+def csv_iterator(data_path, ngrams, vocab, epochs=1, is_worker=False):
+    tokenizer = get_tokenizer("basic_english")
+    with io.open(data_path, encoding="utf8") as f:
+        f.seek(0, 2)
```

Comment:
This part could be pretty confusing if not commented

Reply:
Agreed, we'll work through all examples and add lots of comments.